File: packages/google_mlkit_selfie_segmentation/ios/Classes/GoogleMlKitSelfieSegmentationPlugin.m
#import "GoogleMlKitSelfieSegmentationPlugin.h"
#import <MLKitSegmentationSelfie/MLKitSegmentationSelfie.h>
#import <MLKitSegmentationCommon/MLKitSegmentationCommon.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_selfie_segmenter"
#define startSelfieSegmenter @"vision#startSelfieSegmenter"
#define closeSelfieSegmenter @"vision#closeSelfieSegmenter"

@implementation GoogleMlKitSelfieSegmentationPlugin {
    NSMutableDictionary *instances;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitSelfieSegmentationPlugin* instance = [[GoogleMlKitSelfieSegmentationPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startSelfieSegmenter]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closeSelfieSegmenter]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (MLKSegmenter*)initialize:(FlutterMethodCall *)call {
    BOOL isStream = [[call.arguments objectForKey:@"isStream"] boolValue];
    BOOL enableRawSizeMask = [[call.arguments objectForKey:@"enableRawSizeMask"] boolValue];
    
    MLKSelfieSegmenterOptions *options = [[MLKSelfieSegmenterOptions alloc] init];
    options.segmenterMode = isStream ? MLKSegmenterModeStream : MLKSegmenterModeSingleImage;
    options.shouldEnableRawSizeMask = enableRawSizeMask;
    
    return [MLKSegmenter segmenterWithOptions:options];
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    MLKVisionImage *image = [MLKVisionImage visionImageFromData:call.arguments[@"imageData"]];
    
    NSString *uid = call.arguments[@"id"];
    MLKSegmenter *segmenter = [instances objectForKey:uid];
    if (segmenter == NULL) {
        segmenter = [self initialize:call];
        instances[uid] = segmenter;
    }
    
    [segmenter processImage:image
                 completion:^(MLKSegmentationMask * _Nullable mask,
                              NSError * _Nullable error) {
        if (error) {
            result(getFlutterError(error));
            return;
        } else if (mask == NULL) {
            result(NULL);
            return;
        }
        
        size_t width = CVPixelBufferGetWidth(mask.buffer);
        size_t height = CVPixelBufferGetHeight(mask.buffer);
        
        CVPixelBufferLockBaseAddress(mask.buffer, kCVPixelBufferLock_ReadOnly);
        size_t maskBytesPerRow = CVPixelBufferGetBytesPerRow(mask.buffer);
        float *maskAddress = (float *)CVPixelBufferGetBaseAddress(mask.buffer);
        
        NSMutableArray *confidences = [NSMutableArray array];
        for (int row = 0; row < height; ++row) {
            for (int col = 0; col < width; ++col) {
                // Gets the confidence of the pixel in the mask being in the foreground.
                float confidence = maskAddress[col];
                [confidences addObject:@(confidence)];
            }
            maskAddress += maskBytesPerRow / sizeof(float);
        }
        
        NSMutableDictionary *dictionary = [NSMutableDictionary new];
        dictionary[@"width"] = @(width);
        dictionary[@"height"] = @(height);
        dictionary[@"confidences"] = confidences;
        
        result(dictionary);
    }];
}

@end

File: packages/google_mlkit_selfie_segmentation/ios/google_mlkit_selfie_segmentation/Sources/google_mlkit_selfie_segmentation/GoogleMlKitSelfieSegmentationPlugin.swift
import Flutter
import MLKitVision
import MLKitSegmentationSelfie
import MLKitSegmentationCommon
import google_mlkit_commons

@objc
public class GoogleMlKitSelfieSegmentationPlugin: NSObject, FlutterPlugin {
  private var instances: [String: Segmenter] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_selfie_segmenter",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitSelfieSegmentationPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "vision#startSelfieSegmenter":
      handleDetection(call: call, result: result)
    case "vision#closeSelfieSegmenter":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initialize(call: FlutterMethodCall) -> Segmenter? {
    guard let args = call.arguments as? [String: Any] else { return nil }
    let isStream = (args["isStream"] as? NSNumber)?.boolValue ?? false
    let enableRawSizeMask = (args["enableRawSizeMask"] as? NSNumber)?.boolValue ?? false
    let options = SelfieSegmenterOptions()
    options.segmenterMode = isStream ? .stream : .singleImage
    options.shouldEnableRawSizeMask = enableRawSizeMask
    return Segmenter.segmenter(options: options)
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let imageData = args["imageData"] as? [String: Any],
          let uid = args["id"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }
    guard let image = VisionImage.visionImage(from: imageData) else {
      result(FlutterError(code: "invalid_image", message: "Invalid or missing image data", details: nil))
      return
    }

    let segmenter: Segmenter
    if let existing = instances[uid] {
      segmenter = existing
    } else {
      guard let newSegmenter = initialize(call: call) else {
        result(FlutterError(code: "invalid_args", message: "Invalid options", details: nil))
        return
      }
      segmenter = newSegmenter
      instances[uid] = segmenter
    }

    segmenter.process(image) { mask, error in
      if let error = error as NSError? {
        result(FlutterError(code: "Error \(error.code)", message: error.domain, details: error.localizedDescription))
        return
      }
      guard let mask = mask else {
        result(nil)
        return
      }
      let width = CVPixelBufferGetWidth(mask.buffer)
      let height = CVPixelBufferGetHeight(mask.buffer)
      CVPixelBufferLockBaseAddress(mask.buffer, .readOnly)
      let maskBytesPerRow = CVPixelBufferGetBytesPerRow(mask.buffer)
      guard let baseAddress = CVPixelBufferGetBaseAddress(mask.buffer) else {
        CVPixelBufferUnlockBaseAddress(mask.buffer, .readOnly)
        result(FlutterError(code: "error", message: "Failed to get pixel buffer base address", details: nil))
        return
      }
      let maskAddress = baseAddress.assumingMemoryBound(to: Float.self)
      var confidences: [NSNumber] = []
      var rowAddress = maskAddress
      for _ in 0..<height {
        for col in 0..<Int(width) {
          confidences.append(NSNumber(value: rowAddress[col]))
        }
        rowAddress = rowAddress.advanced(by: maskBytesPerRow / MemoryLayout<Float>.size)
      }
      CVPixelBufferUnlockBaseAddress(mask.buffer, .readOnly)
      result([
        "width": width,
        "height": height,
        "confidences": confidences
      ])
    }
  }
}
